### PR TITLE
Add new state

### LINKED ISSUE #190: Add new state
**Is your feature request related to a problem? Please describe.**
Related to #105.

**Describe the solution you'd like**
A new state is to be implemented to keep track of the number of active assets in the network, thereby allowing us to ensure that at the time of block proposal, all the active assets are being fulfilled in the proposed block. In this new state, the proposer of the 1st valid block will have to claim the block reward by confirming its proposed block. It is during this state where the number of active assets are also updated for the next epoch.

### PR DESCRIPTION
**Pull request by SamAg19**
Fixes #190

### CODE REVIEW

File: contracts/Core/StateManager.sol [hunk 1]
```diff
@@ -29,7 +29,7 @@ contract StateManager is Constants {
     }
 
     function getState(uint32 epochLength) public view returns (State) {
-        uint8 state = uint8(((block.number) / (epochLength / NUM_STATES)));
+        uint32 state = uint32(((block.number) / (epochLength / NUM_STATES)));
```
Comment (SkandaBhat) [inline contracts/Core/StateManager.sol:32]: This can be uint8. Any specific reason it was changed to uint32?

**Review by SkandaBhat — CHANGES_REQUESTED** [review]

File: contracts/Core/StateManager.sol (hunk 1 above)
Comment (SamAg19) [reply contracts/Core/StateManager.sol:32]: With the increase in block number, the value of state also increase to the point where it surpasses 255 and starts again from 0. This change was causing the getState to return the wrong state. Essentially with every 15300 blocks that are mined, the state also resets to 0. Hence, changed it to uint32 to be safe and not to encounter such issues again

Comment (hrishikeshio) [reply contracts/Core/StateManager.sol:32]: the calculation inside brackets happens at uint256. after calculation value is between (0,4). the uint8() converts it to uint8. How can it overflow?

Comment (SkandaBhat) [reply contracts/Core/StateManager.sol:32]: Sameer might be correct.

Dry run:
block.number: 15300
epochLength: 300
NUM_STATES: 5

uint8 state = uint8 (15300/(300/5))
state = 255
return 255%5 = 0

Anything above this and state will start overflowing. We can perform modulus and return without storing state maybe.

Comment (hrishikeshio) [reply contracts/Core/StateManager.sol:32]: Oh I see. The %5 needs to be done in the bracket then. I thought that was being done already.

uint32(((block.number) / (epochLength / NUM_STATES))%5);

Comment (SkandaBhat) [reply contracts/Core/StateManager.sol:32]: Ya this works. cc @SamAg19

Comment (SamAg19) [reply contracts/Core/StateManager.sol]: Done, implemented the change

**Review by hrishikeshio — APPROVED** [review]

**Review by SkandaBhat — APPROVED** [review]

**Review by abhishekvispute — APPROVED** [review]

**Pull request merged**